v3iofs/fs.py

```diff
@@ -118,58 +118,64 @@ def ls(self, path, detail=True, marker=None, **kwargs):
         container, path = split_container(path)
         if not container:
             return self._list_containers(detail)
-
-        resp = self._client.get_container_contents(
-            container=container,
-            path=path,
-            get_all_attributes=True,
-            raise_for_status=v3io.dataplane.RaiseForStatus.never,
-            marker=marker,
-        )
-
-        # Ignore 404's here
-        if resp.status_code not in {200, 404}:
-            raise Exception(
-                f'{resp.status_code} received while accessing {path!r}')
-
-        out = (
-                _resp_dirs(resp, container, detail) +
-                _resp_files(resp, container, detail)
-        )
-
-        if not marker:
-            # first time in
-            if not _has_data(resp):
-                return [self._ls_file(container, path, detail)]
-            if not out:
-                raise FileNotFoundError(f'{full_path!r} not found')
-
-        if hasattr(resp.output, 'next_marker') and resp.output.next_marker:
-            marker = resp.output.next_marker
-            out.extend(self.ls(full_path, detail, marker))
-        return out
+        ext_out = []
+
+        while True:
+            resp = self._client.get_container_contents(
+                container=container,
+                path=path,
+                get_all_attributes=True,
+                raise_for_status=v3io.dataplane.RaiseForStatus.never,
+                marker=marker,
+            )
+
+            # Ignore 404's here
+            if resp.status_code not in {200, 404}:
+                raise Exception(
+                    f'{resp.status_code} received while accessing {path!r}')
+
+            out = (
+                    _resp_dirs(resp, container, detail) +
+                    _resp_files(resp, container, detail)
+            )
+
+            if not marker:
+                # first time in
+                if not _has_data(resp):
+                    return [self._ls_file(container, path, detail)]
+                if not out:
+                    raise FileNotFoundError(f'{full_path!r} not found')
+
+            ext_out.extend(out)
+            if hasattr(resp.output, 'next_marker') and resp.output.next_marker:
+                marker = resp.output.next_marker
+            else:
+                break
+        return ext_out
 
     def _ls_file(self, container, path, detail, marker=None):
```

Comment:
unrelated to your change, looking at the logic of _ls_file() it'm implementation is highly inefficient - running get_container_contents on the parent dir and looking for the file.
instead, it's enough to use v3io-fs head: https://github.com/v3io/v3io-py/blob/2b6541611286e1bd5c151d7af66e729d68497ac8/v3io/dataplane/object.py#L14

Reply:
requires more attrs that are not available in v3io-fs head.
for now skipping this optimization - we'll get back to it once we have nginx get-attr as part of ML-2218